rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -297,90 +451,87 @@ public static FlowNode getFlowEndNode(FlowExecution execution) {
-        final List<FlowNode> nodes = new ArrayList<FlowNode>();
+    // Throws ConcurrentModificationException if FlowGraph changes under the iterator
+    public static List<FlowNode> getIdSortedExecutionNodeList(FlowExecution execution) throws ConcurrentModificationException {
+        if (execution == null || execution.getCurrentHeads().isEmpty()) {
+            return Collections.EMPTY_LIST;
+        }
 
-        FlowGraphWalker walker = new FlowGraphWalker(parentNode.getExecution());
-        for (FlowNode node : walker) {
-            if (node.getParents().contains(parentNode) && !nodes.contains(node)) {
-                nodes.add(node);
+        String executionUrl = null;
+        boolean isCacheable = false;
+        try {
+            executionUrl = execution.getUrl();
+        } catch (IOException ioe) {
+            LOGGER.severe("Can't get execution url for execution, IOException!");
+        }
+        Cache<String,List<FlowNode>> cache = CacheExtension.all().get(0).getExecutionCache();
+        if (executionUrl != null) {
+            List<FlowNode> sortedList = cache.getIfPresent(executionUrl);
+            if (sortedList != null) {
+                return sortedList;
             }
         }
-        sortNodesById(nodes);
-
-        return nodes;
-    }
-
-    public static List<FlowNode> getIdSortedExecutionNodeList(FlowExecution execution) {
-        return getIdSortedExecutionNodeList(execution.getCurrentHeads());
-    }
 
-    public static List<FlowNode> getIdSortedExecutionNodeList(List<FlowNode> nodeList) {
-        if (nodeList == null || nodeList.isEmpty()) {
-            return Collections.EMPTY_LIST;
+        if (isNotPartOfRunningBuild(execution)) {
+            isCacheable = true;
         }
 
-        FlowNodeListCacheAction cacheAction = FlowNodeListCacheAction.get(nodeList.get(0));
-        if (cacheAction.idSortedNodeList == null) {
-            List<FlowNode> unsortedNodes = getUnsortedSortedExecutionNodeList(nodeList);
-            cacheAction.idSortedNodeList = sortNodesById(unsortedNodes);
+        // Not in cache or can't cache
+        FlowGraphWalker walker = new FlowGraphWalker(execution);
+        ArrayList<FlowNode> nodes = new ArrayList<FlowNode>();
+        for (FlowNode node : walker) {
+            nodes.add(node);
         }
+        sortNodesById(nodes);
 
-        return cacheAction.idSortedNodeList;
-    }
-
-    public static List<FlowNode> getUnsortedSortedExecutionNodeList(FlowExecution execution) {
-        return getUnsortedSortedExecutionNodeList(execution.getCurrentHeads());
+        if (isCacheable && executionUrl != null) {
+            cache.put(executionUrl, nodes);
+        }
+        return nodes;
     }
 
-    static List<FlowNode> getUnsortedSortedExecutionNodeList(List<FlowNode> nodeList) {
-        if (nodeList == null || nodeList.isEmpty()) {
-            return Collections.EMPTY_LIST;
+    @Extension
```

Comment:
Nothing important, just that `@Extension` on top of the javadoc is confusing, this:

```java
/** 
 * This is used to cover an obscure case where a WorkflowJob is renamed BUT
 * a previous WorkflowJob existed with cached execution data.
 * Otherwise the previous job's cached data would be returned.
 */
@Extension
public static class RenameHandler extends ItemListener {
```

would be more readable.